discovery/src/main/java/org/ethereum/beacon/discovery/enr/NodeRecord.java

```diff
@@ -0,0 +1,192 @@
+  public static String FIELD_IP_V6 = "ip6";
+  // IPv6-specific TCP port
+  public static String FIELD_TCP_V6 = "tcp6";
+  // IPv6-specific UDP port
+  public static String FIELD_UDP_V6 = "udp6";
+
+  private UInt64 seq;
+  // Signature
+  private BytesValue signature;
+  // optional fields
+  private Map<String, Object> fields = new HashMap<>();
+
+  private EnrSchemeInterpreter enrSchemeInterpreter;
+
+  private NodeRecord(EnrSchemeInterpreter enrSchemeInterpreter, UInt64 seq, BytesValue signature) {
+    this.seq = seq;
+    this.signature = signature;
+    this.enrSchemeInterpreter = enrSchemeInterpreter;
+  }
+
+  private NodeRecord() {}
+
+  public static NodeRecord fromValues(
+      EnrSchemeInterpreter enrSchemeInterpreter,
+      UInt64 seq,
+      BytesValue signature,
+      List<Pair<String, Object>> fieldKeyPairs) {
+    NodeRecord nodeRecord = new NodeRecord(enrSchemeInterpreter, seq, signature);
+    fieldKeyPairs.forEach(objects -> nodeRecord.set(objects.getValue0(), objects.getValue1()));
+    return nodeRecord;
+  }
+
+  public static NodeRecord fromRawFields(
+      EnrSchemeInterpreter enrSchemeInterpreter,
+      UInt64 seq,
+      BytesValue signature,
+      List<RlpType> rawFields) {
+    NodeRecord nodeRecord = new NodeRecord(enrSchemeInterpreter, seq, signature);
+    for (int i = 0; i < rawFields.size(); i += 2) {
+      String key = new String(((RlpString) rawFields.get(i)).getBytes());
+      nodeRecord.set(key, enrSchemeInterpreter.decode(key, (RlpString) rawFields.get(i + 1)));
+    }
+    return nodeRecord;
+  }
+
+  public String asBase64() {
+    return new String(Base64.getUrlEncoder().encode(serialize().extractArray()));
+  }
+
+  public EnrScheme getIdentityScheme() {
+    return enrSchemeInterpreter.getScheme();
+  }
+
+  public void set(String key, Object value) {
+    fields.put(key, value);
+  }
+
+  public Object get(String key) {
+    return fields.get(key);
+  }
+
+  public UInt64 getSeq() {
+    return seq;
+  }
+
+  public void setSeq(UInt64 seq) {
+    this.seq = seq;
+  }
+
+  public BytesValue getSignature() {
+    return signature;
+  }
+
+  public void setSignature(BytesValue signature) {
+    this.signature = signature;
+  }
+
+  public Set<String> getKeys() {
+    return new HashSet<>(fields.keySet());
+  }
+
+  public Object getKey(String key) {
+    return fields.get(key);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    NodeRecord that = (NodeRecord) o;
+    return Objects.equal(seq, that.seq)
+        && Objects.equal(signature, that.signature)
+        && Objects.equal(fields, that.fields);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(seq, signature, fields);
+  }
+
+  public void verify() {
+    enrSchemeInterpreter.verify(this);
+  }
+
+  public RlpList asRlp() {
+    return asRlp(true);
+  }
+
+  public RlpList asRlp(boolean withSignature) {
```

Comment:
I think to increase readability it's better to create two shortcuts `asRlp` and `asRlpNoSignature`. Then reader won't need to check what's this boolean parameter for.